File: python/src/curious/project.py
from __future__ import annotations

import os
import re
from dataclasses import dataclass
from pathlib import Path

DEFAULT_SPEC_REL = "spec/SPEC.md"
CONFIG_FILENAME = "curious.config.json"
AGENTS_FILENAME = "AGENTS.md"
README_FILENAME = "README.md"


@dataclass
class AgentsDocument:
    path: Path
    rel_path: str
    content: str


@dataclass
class DiscoveredProject:
    project_root: Path
    spec_path: Path
    has_spec: bool
    config_path: Path | None = None


def path_exists(path: Path) -> bool:
    return path.is_file() or path.is_dir()


def resolve_project_at_directory_sync(directory: str | Path) -> DiscoveredProject:
    project_root = Path(directory).resolve()
    spec_path = project_root / DEFAULT_SPEC_REL
    config_path = project_root / CONFIG_FILENAME
    return DiscoveredProject(
        project_root=project_root,
        spec_path=spec_path,
        has_spec=spec_path.is_file(),
        config_path=config_path if config_path.is_file() else None,
    )


def discover_project_in_parents(start_dir: str | Path) -> DiscoveredProject | None:
    directory = Path(start_dir).resolve()
    while True:
        resolved = resolve_project_at_directory_sync(directory)
        if resolved.has_spec:
            return resolved
        if (directory / README_FILENAME).is_file():
            return resolved
        parent = directory.parent
        if parent == directory:
            return None
        directory = parent


def should_discover_parents() -> bool:
    return os.environ.get("CURIOUS_DISCOVER") == "parents"


def load_agents_document(
    project_root: Path, agent_cwd: Path
) -> AgentsDocument | None:
    for candidate in (
        project_root / AGENTS_FILENAME,
        agent_cwd / AGENTS_FILENAME,
    ):
        if candidate.is_file():
            return AgentsDocument(
                path=candidate,
                rel_path=relative_to_root(project_root, candidate),
                content=candidate.read_text(encoding="utf-8"),
            )
    return None


def relative_to_root(project_root: Path, absolute_path: Path) -> str:
    rel = absolute_path.relative_to(project_root)
    return str(rel) if str(rel) != "." else "."


def slug_from_path(project_root: Path) -> str:
    base = project_root.name
    slug = re.sub(r"[^a-z0-9]+", "-", base.lower()).strip("-")
    return slug or "project"

File: src/project.ts
import { access, readFile } from "node:fs/promises";
import path from "node:path";

export interface AgentsDocument {
  path: string;
  relPath: string;
  content: string;
}

export const DEFAULT_SPEC_REL = "spec/SPEC.md";
export const CONFIG_FILENAME = "curious.config.json";
export const AGENTS_FILENAME = "AGENTS.md";
export const README_FILENAME = "README.md";

export interface DiscoveredProject {
  projectRoot: string;
  specPath: string;
  hasSpec: boolean;
  configPath?: string;
}

export async function pathExists(filePath: string): Promise<boolean> {
  try {
    await access(filePath);
    return true;
  } catch {
    return false;
  }
}

/**
 * Project root is the directory you run `curious` from (or CURIOUS_CWD).
 * Spec is always `<projectRoot>/spec/SPEC.md`.
 */
export async function resolveProjectAtDirectory(
  directory: string,
): Promise<DiscoveredProject> {
  const projectRoot = path.resolve(directory);
  const specPath = path.join(projectRoot, DEFAULT_SPEC_REL);
  const configPath = path.join(projectRoot, CONFIG_FILENAME);

  return {
    projectRoot,
    specPath,
    hasSpec: await pathExists(specPath),
    configPath: (await pathExists(configPath)) ? configPath : undefined,
  };
}

/**
 * Optional: walk parent directories (set CURIOUS_DISCOVER=parents).
 */
export async function discoverProjectInParents(
  startDir: string,
): Promise<DiscoveredProject | null> {
  let directory = path.resolve(startDir);

  for (;;) {
    const resolved = await resolveProjectAtDirectory(directory);
    if (resolved.hasSpec) {
      return resolved;
    }
    if (await pathExists(path.join(directory, README_FILENAME))) {
      return resolved;
    }

    const parent = path.dirname(directory);
    if (parent === directory) {
      return null;
    }
    directory = parent;
  }
}

export function shouldDiscoverParents(): boolean {
  return process.env.CURIOUS_DISCOVER === "parents";
}

/**
 * Load AGENTS.md from project root, then agent cwd. Returns full file content.
 */
export async function loadAgentsDocument(
  projectRoot: string,
  agentCwd: string,
): Promise<AgentsDocument | undefined> {
  const candidates = [
    path.join(projectRoot, AGENTS_FILENAME),
    path.join(agentCwd, AGENTS_FILENAME),
  ];

  for (const absolutePath of candidates) {
    if (!(await pathExists(absolutePath))) {
      continue;
    }

    const content = await readFile(absolutePath, "utf8");
    return {
      path: absolutePath,
      relPath: relativeToRoot(projectRoot, absolutePath),
      content,
    };
  }

  return undefined;
}

export function relativeToRoot(projectRoot: string, absolutePath: string): string {
  const relative = path.relative(projectRoot, absolutePath);
  return relative || ".";
}

export function slugFromPath(projectRoot: string): string {
  const base = path.basename(projectRoot);
  const slug = base
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
  return slug || "project";
}
